components/brave_wallet/browser/solana_data_decoder_utils.cc

```diff
@@ -0,0 +1,204 @@
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<uint64_t> DecodeUint64(const std::vector<uint8_t>& input,
+                                      size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint64_t)) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  base::span<const uint8_t> s =
+      base::make_span(input.begin() + offset, sizeof(uint64_t));
+  uint64_t uint64_le = *reinterpret_cast<const uint64_t*>(s.data());
+
+  offset += sizeof(uint64_t);
+
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
+  return uint64_le;
+#else
+  return base::ByteSwap(uint64_le);
+#endif
+}
+
+absl::optional<std::string> DecodeUint64String(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint64(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<std::string> DecodePublicKey(const std::vector<uint8_t>& input,
+                                            size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < kSolanaPubkeySize)
+    return absl::nullopt;
+
+  offset += kSolanaPubkeySize;
+  return Base58Encode(std::vector<uint8_t>(
+      input.begin() + offset - kSolanaPubkeySize, input.begin() + offset));
+}
+
+absl::optional<std::string> DecodeOptionalPublicKey(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  if (offset == input.size()) {
+    return absl::nullopt;
+  }
+
+  // First byte is 0 or 1 to indicate if public key is passed.
+  // And the rest bytes are the actual public key.
+  if (input[offset] == 0) {
+    offset++;
+    return "";  // No public key is passed.
+  } else if (input[offset] == 1) {
+    offset++;
+    return DecodePublicKey(input, offset);
+  } else {
+    return absl::nullopt;
+  }
+}
+
+// bincode::serialize uses two u32 together for the string length and a byte
+// array for the actual strings. The first u32 represents the lower bytes of
+// the length, the second represents the upper bytes. The upper bytes will have
+// non-zero value only when the length exceeds the maximum of u32.
+// We currently cap the length here to be the max size of std::string
+// on 32 bit systems, it's safe to do so because currently we don't expect any
+// valid cases would have strings larger than it.
+absl::optional<std::string> DecodeString(const std::vector<uint8_t>& input,
+                                         size_t& offset) {
+  auto len_lower = DecodeUint32(input, offset);
+  if (!len_lower || *len_lower > kMaxStringSize32Bit)
+    return absl::nullopt;
+  auto len_upper = DecodeUint32(input, offset);
+  if (!len_upper || *len_upper != 0) {  // Non-zero means len exceeds u32 max.
+    return absl::nullopt;
+  }
+
+  if (offset + *len_lower > input.size())
+    return absl::nullopt;
+
+  offset += *len_lower;
+  return std::string(reinterpret_cast<const char*>(&input[offset - *len_lower]),
```

Comment:
you don't really need the reinterpret magic, just something like this should work
`std::string(input.begin() + offset - *len_lower, input.begin() + offset)`